rdtools/availability.py

```diff
@@ -0,0 +1,576 @@
+        df_outages['type'] = np.where(
+            df_outages['energy_actual'] < df_outages['ci_lower'],
+            'real',
+            'comms')
+        df_outages.loc[df_outages['energy_actual'].isnull(), 'type'] = 'unknown'
+        df_outages['loss'] = np.where(
+            df_outages['type'] == 'real',
+            df_outages['energy_expected'] - df_outages['energy_actual'],
+            0)
+        df_outages.loc[df_outages['type'] == 'unknown', 'loss'] = np.nan
+
+        self.outage_info = df_outages
+
+        # generate a best-guess timeseries loss for the full outages by
+        # scaling the expected power signal to match the actual
+        lost_power_full = pd.Series(0, index=self.loss_subsystem.index)
+        expected_power = self.power_expected
+        corrected_cumulative_energy = self.energy_cumulative.copy()
+        for i, row in self.outage_info.iterrows():
+            start = row['start']
+            end = row['end']
+            subset = expected_power.loc[start:end].copy()
+            subset_energy = rdtools.energy_from_power(subset)
+            loss_fill = subset * row['loss'] / subset_energy.sum()
+            lost_power_full.loc[subset.index] += loss_fill
+
+            # fill in the cumulative meter during the outages, again using
+            # the expected energy signal, but rescaled to match actual
+            # production this time:
+            production_fill = subset_energy.cumsum()
+            production_fill *= row['energy_actual'] / subset_energy.sum()
+            corrected_segment = row['energy_start'] + production_fill
+            corrected_cumulative_energy.loc[start:end] = corrected_segment
+
+        self.energy_cumulative_corrected = corrected_cumulative_energy
+        self.loss_system = lost_power_full
+
+    def _combine_losses(self, rollup_period='M'):
+        """
+        Combine subsystem and system losses.
+
+        Sets the `loss_total` and `results` attributes.
+
+        Parameters
+        ----------
+        rollup_period : pandas offset string, default 'M'
+            The period on which to roll up losses and calculate availability.
+        """
+        self.loss_total = self.loss_system + self.loss_subsystem
+
+        # calculate actual production based on corrected cumulative meter
+        cumulative_energy = self.energy_cumulative_corrected
+        resampled_cumulative = cumulative_energy.resample(rollup_period)
+        actual_production = (
+            resampled_cumulative.last() - resampled_cumulative.first()
+        )
+
+        lost_production = rdtools.energy_from_power(self.loss_total)
+        df = pd.DataFrame({
+            'lost_production': lost_production.resample(rollup_period).sum(),
+            'actual_production': actual_production,
+        })
+        loss_plus_actual = df['lost_production'] + df['actual_production']
+        df['availability'] = 1 - df['lost_production'] / loss_plus_actual
+        self.results = df.tz_localize(None)
```

Comment:
What's the reasoning for this line? Why not just use the same timezone as the original input data?

Reply:
That line is a holdover from when this was internal code.  The reasoning was that, for monthly results, the timezone is not meaningful and was just getting in the way of merging the roll-up datasets with different timezones.  The merging aspect is less relevant here, and I still think it doesn't really make sense for these timestamps to have timezones, but maybe that's too presumptuous and we should let the user drop the tz afterwards if they want to.

Reply:
I would err on the side of not stripping out the time zone information. I'm trying to think generally. For example, folks could be interested in their hourly availability, in which case time zone information may be desirable to understand what is going on across the fleet.